### ISSUE TITLE
`Prelude` module clashes with `base`

### ISSUE THREAD

**Issue by sjakobi** [haskell-nix/hnix#994]
v0.15.0 exposes a new `Prelude` module that makes it difficult to use `hnix`, e.g.:

```
src/Dhall/Nix.hs:87:8: error:
    Ambiguous module name ‘Prelude’:
      it was found in multiple packages: base-4.14.3.0 hnix-0.15.0
   |
87 | module Dhall.Nix (
   |        ^^^^^^^^^

```

Could this `Prelude` be made internal or alternatively be exposed under a different name, say, `Nix.Prelude`?

**Comment by Anton-Latukha** [comment]
Woops.

Oh, yes.
Was thinking about it while making it.
Would fix it in nearest time.

Probably making it internal would just work.

Actually I was more surprised/impressed that what I had in mind - I made it work. The setup combines the mixing loaded `relude`s `Prelude` & overshadows it with project local `Utils` into 1 `Prelude` module. So no imports are needed, and name conflicts are resolved.

But forgot to resolve the main outside naming conflict.

**Comment by Anton-Latukha** [comment]
Ye....

Cabal works only this way and does not allow to not export the `Prelude` :man_shrugging: ... It is strange... If I put it in, for example, `other-modules` - project modules no longer see `Prelude` module.

Probably would need to go `NoImplicitPrelude` route & import it as `Nix.Prelude` everywhere  ...

Today is late, would do it tomorrow.

I made a minor release beforehand, so if some stuff like this happens - there is a backup option.

**Comment by sjakobi** [comment]
> If I put it in, for example, `other-modules` - project modules no longer see `Prelude` module.

Sounds like a `Cabal` bug to me. Might be worth reporting.

In any case this isn't an urgent issue for me. `dhall-nix*` can simply continue using v0.14 for now.

**Comment by Anton-Latukha** [comment]
But as an active HNix maintainer - it is one for me. It is a disservice to make Cabal promote a broken release.

The proper thing in these cases - is to demote that version right away.

I going to make a new release & then open a Cabal report ... ("So many houses, so little time")

**Comment by Anton-Latukha** [comment]
Deprecated the `0.15.0`.

**Comment by kquick** [comment]
The `0.15.0` version is still being chosen by `cabal` for build plans, so this is forcing all packages using `hnix` to add an upper-bound constraint of `< 0.13`.  Is there a timeframe for a fix (and what is the fix.. rolling back to `Prelude`?)?

**Comment by sjakobi** [comment]
> The `0.15.0` version is still being chosen by `cabal` for build plans

Can you give an example?

It might be helpful to add an impossible constraint like `base < 0` to `v0.15.0`.

**Comment by Anton-Latukha** [comment]
If `0.15` is deprecated why particularly `< 0.13` then, `0.12` would be a version like 1.5 years ago.

> Is there a timeframe.

Well. Nix 2.3.15 released with a breakage without documentation & then Nix 2.4 was released with regressions in "stable" features & without increasing the `languageVersion` shipped undocumented changes in the handling of the language. That is not just problem for me, but central Nixpkgs maintainers started & gather signings in the declaration: https://discourse.nixos.org/t/nix-2-4-and-what-s-next/16257 to stop these sudden undocumented top-down breakages

 & HNix has & required to follow Nix bug2bug if that is needed & so uses Nix for golden tests. From the tests enabled there is currently breakage in 1 escaping mechanism moment that Nix changed. Because it is not documented, I hunted through the Nix git logs & found no explicit place where escaping was changed, & read through a bunch of C++ at the best of my possibilities, tried to bisect the issue, but you get the picture. The is a need for someone who knows C++ better than me.
 
I tried to adjust the HNix behavior in https://github.com/haskell-nix/hnix/pull/1009 but failed. & then I decided that if nobody really cares for this - I better do some work in HLS then, which I do currently.

Seems like HNix does the representation of this by running the `iterNValue` directly. I'd loved to see the C++ code - to know in what part of the process the escaping changes.

**Comment by Anton-Latukha** [comment]
& I'd personally put those tests as broken for a time & worked on other things. Because there is a number of new tests to take from Nix that do work & other ones are easier to do for me.

**Comment by Anton-Latukha** [comment]
Set a suggested impossible bound (`< 0`) on the `0.15`.

**Comment by kquick** [comment]
The main project is not public, but here's a simple reproduction.  First, the basic setup of a method to test this:

```
$ mkdir hnix-test
$ cd hnix-test
$ cabal --version
cabal-install version 3.6.0.0
compiled using version 3.6.1.0 of the Cabal library
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.10.7
$ cabal init
...
$ cabal run
...
Hello, Haskell!
```

Then I add hnix as a dependency.  With no constraints, cabal selects 0.14.0.5, but alas, this does not build:

```
$ sed -i -e '/build-depends:/s/build-depends: .*/build-depends: base, hnix/' hnix-test.cabal
$ cabal run
Build profile: -w ghc-8.10.7 -O1
...
 - hnix-0.14.0.5 (lib) (requires download & build)
 - megaparsec-9.1.0 (lib)
...
Starting    hnix-0.14.0.5 (lib)
Building    hnix-0.14.0.5 (lib)

Failed to build hnix-0.14.0.5
...
[32 of 50] Compiling Paths_hnix     ( dist/build/autogen/Paths_hnix.hs, dist/build/Paths_hnix.o, dist/build/Paths_hnix.dyn_o )

dist/build/autogen/Paths_hnix.hs:66:22: error:
    * 'last' works with 'NonEmpty', not ordinary lists.
      ...
66|   | isPathSeparator (last dir) = dir ++ fname
```

It works fine for hnix 0.12.0.1:

```
$ cabal run --constraint 'hnix < 0.13'
...
 - hnix-0.12.0.1 (lib) (requires download & build)
 - megaparsec-9.0.1 (lib)
...
Hello, Haskell!
```
Trying 0.13.1 fails:

```
$ cabal run --constraint 'hnix < 0.14'
...
 - hnix-0.13.1 (lib) (requires download & build)
 - megaparsec-9.0.1 (lib)
...
[same error as 0.14: 'last' works with 'NonEmpty', not ordinary lists. at line 66 of Paths_hnix.hs]
```

Interestingly, cabal was happy to select 0.14.0.5 when no other constraints were specified, but when adding a megaparsec dependency  explicitly, things get strange:

```
$ sed -i -e '/build-depends:/s/build-depends: .*/build-depends: base, hnix, megaparsec/' hnix-test.cabal
$ cabal run
...
 - megaparsec-9.2.0 (lib) (requires build)
 - hnix-0.2.1 (lib) (requires build)
 ...
Hello, Haskell!
```

The selection of version 0.2.1 of `hnix` is rather surprising there.  Adding a lower bound constraint on hnix:

```
$ cabal run --constraint 'hnix >= 0.11'
...
 - megaparsec-9.2.0 (lib) (requires build)
 - hnix-0.14.0.2 (lib) (requires build)
...
[same 'last' error]
```

Now I need to add one more dependency to get the final behavior:

```
$ sed -i -e '/build-depends:/s/build-depends: .*/build-depends: base, hnix, megaparsec, servant/' hnix-test.cabal
$ cabal run --constraint 'hnix >= 0.11'
...
 - servant-0.18.3 (lib) (requires build)
 - hnix-0.15.0 (lib) (requires build)
...
Building library for hnix-0.15.0..
[ 1 of 52] Compiling Nix.Utils    ( src/Nix/Utils.hs, dist/build/Nix/Utils.o, dist/build/Nix/Utils.o, dist/build/Nix/Utils.dyn_o )
[ 2 of 52] Compiling Prelude     ( src/Prelude.hs, dist/build/Prelude.o, dist/build/Prelude.dyn_o )
[ 3 of 52] Compiling Paths_hnix  ( dist/build/autogen/Paths_hnix.hs, dist/build/Paths_hnix.o, dist/build/Paths_hnix.dyn_o )

dist/build/autogen/Paths_hnix.hs:66:22: error:
    * 'last' works with 'NonEmpty', not ordinary lists
...
```

Based on the above, it appears that:
 * all of the releases after 0.12.0.1 are failing (due the switch to `relude` ?) and 
 * there are some package combinations that will select hnix 0.15.0 without explicit upper-bounds preventing this

**Comment by Anton-Latukha** [comment]
`0.15` supported `megaparsec` `0.9.2`.

That may be the thing that makes it choose it.

What the hell is `Path_hnix` error, if that module is autogenerated.

I would look into it when I have time & would try to make a release soon, but it is probably no earlier then in a week.

But since I currenly do not know the cause of the syndrome - currently do not know if new release would fix the autogeneration of the module.

**Comment by sjakobi** [comment]
@kquick If you `cabal update`, do you still get build plans with `hnix-0.15`? You shouldn't.

Regarding the weird build failures in `Paths_hnix.hs`, I think you're running into a `cabal` bug. See https://github.com/haskell/cabal/blob/master/release-notes/cabal-install-3.6.2.0.md#significant-changes.

If you update your `cabal` installation, it should be gone.

**Comment by kquick** [comment]
Running a new `cabal update` seems to resolve the selection of 0.15:

```
$ cabal update 'hackage.haskell.org,2021-12-03T15:52:06Z'
$ cabal run --constraint 'hnix >= 0.11'
...
 - hnix-0.15.0
...
$ cabal update 'hackage.haskell.org,2021-12-04T14:47:15Z'
$ cabal run --constraint 'hnix >= 0.11'
...
 - hnix-0.14.0.5
...
```

Also, upgrading cabal from 3.6.0.0 to 3.6.2.0 does fix the `Paths_hnix` compile problem and I can successfully build the 0.14.0.5 version.

Interestingly, without the lower bound constraint and when an unconstrained `megaparsec` is part of the build dependencies, cabal still chooses the very old `hnix-0.2.1` (and version 9.2.0 of `megaparsec`).  Removing `megaparsec` changes the build plan back to 0.14.0.5.  Shrug.

Thank you both, @sjakobi and @Anton-Latukha  for your help on this issue (and your work on hnix).  I now have a better plan for using a much more recent version of hnix than 0.12.

**Comment by Anton-Latukha** [comment]
@kquick Thank you for contributing information & reporting the situation. 

Thank you Simon for help here, he is closer to the center of the Haskell stack & knows the specifics. Do not know your current situation hope you would get a living out of it.

Marking these messages as resolved.

This topic overall is still open, as somebody (I) still needs to unbork the bork.

**Comment by Anton-Latukha** [comment]
Currently resolved in `master`.

**Comment by Anton-Latukha** [comment]
Released `0.16.0`, which uses `NoImplicitPrelude` & prelude is in `Nix.Prelude`.
